main/src/com/google/refine/expr/functions/ToJson.java

```diff
@@ -0,0 +1,87 @@
+/*
+
+Copyright 2010, Google Inc.
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are
+met:
+
+    * Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above
+copyright notice, this list of conditions and the following disclaimer
+in the documentation and/or other materials provided with the
+distribution.
+    * Neither the name of Google Inc. nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,           
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY           
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+package com.google.refine.expr.functions;
+
+import java.time.OffsetDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Properties;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.google.refine.expr.EvalError;
+import com.google.refine.grel.ControlFunctionRegistry;
+import com.google.refine.grel.Function;
+import com.google.refine.util.ParsingUtilities;
+
+public class ToJson implements Function {
+
+    @Override
+    /**
+     * Transfor an array of objects into a json array format string
+     * @param args the user input
+     *        args[0] is an array of objects
+     * @return string representing array of objects in json array format
+     */
+    public Object call(Properties bindings, Object[] args) {
+        String error_message = " accepts an array of objects";
+        Object result = new EvalError(ControlFunctionRegistry.getFunctionName(this) + error_message);
+        if (args.length == 1 && args[0] instanceof Object[]) {
+            Object[] array = (Object[]) args[0];
+            try{
+                result = ParsingUtilities.mapper.writeValueAsString(array);
+                String result_str = (String) result;
+                // add whitespace based on formatting
+                result_str = result_str.replace(",",", ").replace("[","[ ").replace("]"," ]").replace("[  ]","[ ]");
```

Comment:
I think those `replace` calls are unfortunately incorrect in general.
For instance, if you do `toJson(["[,]"])`, you will get `"[\"[ ,  ]\"]"`, which is incorrect. I would remove all of those calls to `replace`. If you want to have a pretty-printed JSON output (which should not be necessary) then Jackson's native features for that should be used.

Reply:
Thanks for pointing out that, I didn't think of that case. I did that to make the function can do what the Expected Behavior shows in issue 4194. I will change it to use Jackson's features. I'm not familiar with that, thanks for your reminder.

Reply:
I think it is fine not to follow the exact formatting given in #4194, the differences are only cosmetic.